cores/esp8266/uart.c

```diff
@@ -283,42 +287,109 @@ uart_resize_rx_buffer(uart_t* uart, size_t new_size)
 size_t
 uart_get_rx_buffer_size(uart_t* uart)
 {
-    return uart && uart->rx_enabled? uart->rx_buffer->size: 0;
+    return uart && uart->rx_enabled ? uart->rx_buffer->size : 0;
 }
 
 
-void ICACHE_RAM_ATTR 
-uart_isr(void * arg)
+void ICACHE_RAM_ATTR
+uart_isr(void* arg)
 {
-    uart_t* uart = (uart_t*)arg;
-    uint32_t usis = USIS(uart->uart_nr);
-
-    if(uart == NULL || !uart->rx_enabled) 
+    // check to confirm we are being called
+    if (arg == s_uartInterruptContext)
     {
-        USIC(uart->uart_nr) = usis;
-        ETS_UART_INTR_DISABLE();
-        return;
+        // Interrupt handler is shared between UART0 & UART1
+        // so we need to check both
+        for (uint8_t uartNum = 0; uartNum < 2; uartNum++)
+        {
+            if (USIS(uartNum))
+            {
+                if (s_uartInterruptContext[uartNum].callback != NULL)
+                {
+                    s_uartInterruptContext[uartNum].callback(s_uartInterruptContext[uartNum].arg);
+                }
+                else
+                {
+                    // No registered handler, so
+                    // clear all interrupts flags (just in case)
+                    USIE(uartNum) = 0;
+                    USIC(uartNum) = 0xffff;
+                }
+            }
+        }
     }
+}
 
-    if(usis & (1 << UIFF))
+void ICACHE_RAM_ATTR
+uart_isrDefault(void * arg)
+{
+    uart_t* uart = (uart_t*)arg;
+    uint32_t usis = USIS(uart->uart_nr);
+
+    if (usis & (1 << UIFF))
```

Comment:
We have ASSERTs that turn off in a normal build?
I would agree with an ASSERT, but I disagree otherwise.   This is an internal handler and not exposed, and all locations include an object.  Param checks should be reserved for the outermost APIs.   If hit such a check, then there is corruption.

Reply:
As I said, _some_ sanity check. An assert is ok by me, as long as it's clear what it checks (i.e.: include comments).

Reply:
BTW, the original check here was invalid.  It would have dereferenced the null on lines above this check, this is why it was removed.